confluent_kafka/schema_registry/json_schema.py

```diff
@@ -0,0 +1,261 @@
+
+    def __call__(self, ctx, obj):
+        """
+        Serializes an object to the Confluent Schema Registry's JSON binary
+        format.
+
+        Args:
+            ctx (SerializationContext): Metadata pertaining to the serialization
+                operation.
+
+            obj (object): object instance to serializes.
+
+
+        Note:
+            None objects are represented as Kafka Null.
+
+        Raises:
+            SerializerError if any error occurs serializing obj
+
+        Returns:
+            bytes: Confluent Schema Registry formatted JSON bytes
+
+        """
+        if obj is None:
+            return None
+
+        subject = self._subject_name_func(ctx, self._schema_name)
+
+        # Check to ensure this schema has been registered under subject_name.
+        if self._auto_register and subject not in self._known_subjects:
+            # The schema name will always be the same. We can't however register
+            # a schema without a subject so we set the schema_id here to handle
+            # the initial registration.
+            self._schema_id = self._registry.register_schema(subject,
+                                                             self._schema)
+            self._known_subjects.add(subject)
+        elif not self._auto_register and subject not in self._known_subjects:
+            registered_schema = self._registry.lookup_schema(subject,
+                                                             self._schema)
+            self._schema_id = registered_schema.schema_id
+            self._known_subjects.add(subject)
+
+        if self.to_dict is not None:
+            value = self.to_dict(ctx, obj)
+        else:
+            value = obj
+
+        try:
+            validate(instance=value, schema=self._parsed_schema)
+        except ValidationError as ve:
+            raise SerializationError(ve.message)
+
+        with _ContextStringIO() as fo:
+            # Write the magic byte and schema ID in network byte order (big endian)
+            fo.write(struct.pack('>bI', _MAGIC_BYTE, self._schema_id))
+            # write the record to the rest of the buffer
+            json.dump(obj, fo, value)
+
+            return fo.getvalue()
+
+
+class JsonDeserializer(Deserializer):
+    """
+    JsonDeserializer decodes bytes written in the Schema Registry
+    JSON format to an object.
+
+    Args:
+        schema_registry_client (SchemaRegistryClient): Confluent Schema Registry
+            client instance.
+
+        schema_str (str): JSON schema declaration.
+
+        from_dict (callable, optional): Callable(SerializationContext, dict) -> object.
+            Converts dict to an instance of some object.
+
+    .. _Schema definition:
+        https://json-schema.org/understanding-json-schema/reference/generic.html
+
+    """
+    __slots__ = ['_schema', '_registry', 'from_dict']
+
+    def __init__(self, schema_registry_client, schema_str, from_dict=None):
```

Comment:
fwiw in .net i didn't even bother integrating with SR on the consume side. but good job..

Reply:
ahh, you don't actually use any of these.

in python you can't have more than on __init__, so i don't think you can nicely extend to add this capability later. not exactly sure what to do. i think probably document that schema_registry_client & schema_str are currently ignored and may be used in the future.

Reply:
yeah I felt a bit weird about this one. I originally validated that the schema fetched matched the provided schema then I saw we don't do that in .NET so I decided to do away with it. Perhaps we should bring it back though.